sdk/iot/core/doc/mqtt_state_machine.md

```diff
@@ -136,3 +138,120 @@ _Example:_
         }
     }
 ```
+
+### Retrying Operations
+
+Retrying operations requires understanding two aspects: error evaluation (did the operation fail, should the operation be retried) and retry timing (how long to delay before retrying the operation). The IoT Embedded Devices SDK is supplying optional APIs for error classification and retry timing.
+
+#### Error policy
+
+The SDK will not handle protocol-level (WebSocket, MQTT, TLS or TCP) errors. The application-developer is expected to classify and handle errors the following way:
+
+- Authentication errors should not be retried.
+- Communication-related errors other than ones security-related should be considered retriable.
+
+Both IoT Hub and Provisioning services will use `MQTT CONNACK` as described in Section 3.2.2.3 of the [MQTT v3 specification](http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Table_3.1_-).
+
+##### IoT Hub Service Errors
+
+APIs using `az_iot_hub_client_status` report service-side errors to the client through the IoT Hub protocols. (At the time of writing, only Twin responses may return errors.)
+
+The following APIs may be used to determine if the status indicates an error and if the operation should be retried:
+
+```C
+az_iot_hub_client_status status = response.status;
+if (az_iot_hub_client_is_success_status(status))
+{
+    // success case 
+}
+else
+{
+    if (az_iot_hub_client_is_retriable_status(status))
+    {
+        // retry
+    }
+    else
+    {
+        // fail
+    }
+}
+```
+
+##### Provisioning Service Errors
+
+Each Provisioning operation will return a status encoded as `az_iot_provisioning_client_status`.
+After the register operation is complete, the `az_iot_provisioning_client_registration_state registration_information` field will contain the `az_iot_provisioning_client_status error_code` that represents the overall status of the registration operation.  

```

Comment:
nit: could `az_iot_provisioning_client_status error_code` be changed to "... contain the `error_code` of type `az_iot_provisioning_client_status`". Reading quickly it might read as one type since a space and underscore can mesh.

Reply:
Removed since @jspaith @JeffreyRichter and I decided to unify error and retry.